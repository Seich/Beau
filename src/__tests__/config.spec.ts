import Config, { parseBeauConfig } from '../config'

const requireg = require('requireg')
requireg.resolving = false

describe('Config', () => {
    it('should load valid config keys', () => {
        const doc = parseBeauConfig(`
            version: 1
            endpoint: http://martianwabbit.com
            shouldntBeAdded: true
        `)

        const config = new Config(doc)
        expect(config.endpoint).toBe(doc.endpoint)
        expect(config.version).toBe(doc.version)
    })

    it('should load requests', () => {
        const doc = parseBeauConfig(`
            endpoint: http://example.com

            GET /profile: get-profile
            GET /posts: get-posts
            GET /posts/1: get-post
            GET /user:
                alias: user
                headers:
                    hello: world
        `)

        const config = new Config(doc)
        expect(Object.keys(config.requests).length).toBe(4)
    })

    it('should set up defaults for all requests', () => {
        const doc = parseBeauConfig(`
            version: 1
            endpoint: 'http://example.com'

            defaults:
                headers:
                    authentication: hello

            GET /posts/1: get-post
            GET /user:
                alias: user
                headers:
                    hello: world
        `)

        const config = new Config(doc)

        expect(config).toMatchSnapshot()
        Object.values(config.requests).forEach((r: any) => {
            expect(r.headers.authentication).toMatch('hello')
        })
    })

    it('should load multiple hosts', () => {
        const doc = parseBeauConfig(`
          version: 1
          endpoint: http://example.org

          defaults:
            headers:
              hello: mars

          GET /e1: e1

          hosts:
            - host: com
              endpoint: http://example.com

              defaults:
                headers:
                  hello: world
                  world: hello

              GET /e2: e2
              GET /posts: posts

            - host: net
              endpoint: http://example.net

              defaults:
                headers:
                  hello: world
                  world: bye

              GET /e3: e3
              GET /posts: posts

            - host: info
              endpoint: http://example.info

              GET /posts: posts
        `)

        let config = new Config(doc)

        expect(config).toMatchSnapshot()
    })

    it('should namespace all aliases within an host', () => {
        const doc = parseBeauConfig(`
            hosts:
              - host: test1
                endpoint: http://example.com
                GET /posts: posts
              - host: test2
                endpoint: http://example.net
                GET /posts: posts
        `)

        let config = new Config(doc)

        expect(config.requests[0].alias).toEqual('test1:posts')
        expect(config.requests[1].alias).toEqual('test2:posts')
    })

    it(`should throw if host doesn't have a host key`, () => {
        const doc = parseBeauConfig(`
            hosts:
              - endpoint: http://example.com
                GET /posts: posts

              - host: test2
                endpoint: http://example.net
                GET /posts: posts
        `)

        expect(() => new Config(doc)).toThrow()
    })

    it(`should merge host settings with global settings`, () => {
        const doc = parseBeauConfig(`
            defaults:
              headers:
                hello: 1

            hosts:
              - host: test
                endpoint: http://example.net
                GET /posts: posts

              - host: test2
                endpoint: http://example.org
                defaults:
                  headers: false

                GET /posts: posts
        `)

        let config = new Config(doc)
        expect(config.requests[0].headers.hello).toBe(1)
    })

    it(`should allow different settings for the same request`, () => {
        const doc = parseBeauConfig(`
      host: https://example.com
      GET /1:
        - alias: req1
          headers:
            request: 1
        - alias: req2
          headers:
            request: 2
    `)

        let config = new Config(doc)
        expect(config.requests.length).toBe(2)
    })
})
