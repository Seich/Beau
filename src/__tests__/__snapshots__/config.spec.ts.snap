// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Config should load multiple hosts 1`] = `
Config {
  "cookiejar": false,
  "defaults": Object {
    "headers": Object {
      "hello": "mars",
    },
  },
  "endpoint": "http://example.org",
  "environment": Object {},
  "host": undefined,
  "hosts": Array [
    Object {
      "GET /e2": "e2",
      "GET /posts": "posts",
      "defaults": Object {
        "headers": Object {
          "hello": "world",
          "world": "hello",
        },
      },
      "endpoint": "http://example.com",
      "host": "com",
    },
    Object {
      "GET /e3": "e3",
      "GET /posts": "posts",
      "defaults": Object {
        "headers": Object {
          "hello": "world",
          "world": "bye",
        },
      },
      "endpoint": "http://example.net",
      "host": "net",
    },
    Object {
      "GET /posts": "posts",
      "endpoint": "http://example.info",
      "host": "info",
    },
  ],
  "plugins": Plugins {
    "autoload": Array [
      "std",
    ],
    "context": Object {},
    "registry": Object {
      "dynamicValues": Array [],
      "postRequestModifiers": Array [],
      "preRequestModifiers": Array [],
    },
  },
  "requests": Array [
    Object {
      "alias": "e1",
      "cookiejar": false,
      "endpoint": "http://example.org",
      "headers": Object {
        "hello": "mars",
      },
      "request": "GET /e1",
    },
    Object {
      "alias": "com:e2",
      "cookiejar": false,
      "endpoint": "http://example.com",
      "headers": Object {
        "hello": "world",
        "world": "hello",
      },
      "request": "GET /e2",
    },
    Object {
      "alias": "com:posts",
      "cookiejar": false,
      "endpoint": "http://example.com",
      "headers": Object {
        "hello": "world",
        "world": "hello",
      },
      "request": "GET /posts",
    },
    Object {
      "alias": "net:e3",
      "cookiejar": false,
      "endpoint": "http://example.net",
      "headers": Object {
        "hello": "world",
        "world": "bye",
      },
      "request": "GET /e3",
    },
    Object {
      "alias": "net:posts",
      "cookiejar": false,
      "endpoint": "http://example.net",
      "headers": Object {
        "hello": "world",
        "world": "bye",
      },
      "request": "GET /posts",
    },
    Object {
      "alias": "info:posts",
      "cookiejar": false,
      "endpoint": "http://example.info",
      "headers": Object {
        "hello": "mars",
      },
      "request": "GET /posts",
    },
  ],
  "version": 1,
}
`;

exports[`Config should set up defaults for all requests 1`] = `
Config {
  "cookiejar": false,
  "defaults": Object {
    "headers": Object {
      "authentication": "hello",
    },
  },
  "endpoint": "http://example.com",
  "environment": Object {},
  "host": undefined,
  "hosts": Array [],
  "plugins": Plugins {
    "autoload": Array [
      "std",
    ],
    "context": Object {},
    "registry": Object {
      "dynamicValues": Array [],
      "postRequestModifiers": Array [],
      "preRequestModifiers": Array [],
    },
  },
  "requests": Array [
    Object {
      "alias": "get-post",
      "cookiejar": false,
      "endpoint": "http://example.com",
      "headers": Object {
        "authentication": "hello",
      },
      "request": "GET /posts/1",
    },
    Object {
      "alias": "user",
      "cookiejar": false,
      "endpoint": "http://example.com",
      "headers": Object {
        "authentication": "hello",
        "hello": "world",
      },
      "request": "GET /user",
    },
  ],
  "version": 1,
}
`;
